source/rust_verify/src/singular.rs

```diff
@@ -309,35 +307,52 @@ impl SingularEncoder {
             format!("{} {} = {}({})", IDEAL_DECL, IDEAL_G, GROEBNER_APPLY, IDEAL_I);
         let reduce_string = format!("{}({}, {})", REDUCE_APPLY, goal, IDEAL_G);
 
-        // ask for acknowledgement
         let res = format!(
-            "{};\n{};\n{};\n{};\nprint(\"{}\");\n",
-            ring_string, ideal_string, ideal_to_groebner, reduce_string, DONE
+            "{};\n{};\n{};\n{};\n",
+            ring_string, ideal_string, ideal_to_groebner, reduce_string
         );
         Ok(res)
     }
 }
 
-pub fn check_singular_valid(
-    context: &mut air::context::Context,
+fn encode_singular_queries(
     command: &Command,
     func_span: &vir::messages::Span,
-    _query_context: QueryContext<'_, '_>,
-) -> ValidityResult {
+    queries: &mut Vec<(String, vir::messages::Message)>,
+) -> Result<(), ValidityResult> {
     let query: Query = if let CommandX::CheckValid(query) = &**command {
         query.clone()
     } else {
         panic!("internal error: integer_ring")
     };
 
-    let decl = query.local.clone();
-    let statement = query.assertion.clone();
-    let stmts: air::ast::Stmts = if let air::ast::StmtX::Block(stmts) = &*statement {
-        stmts.clone()
+    let reqs_enss: air::ast::Stmts =
+        if let air::ast::StmtX::Block(stmts) = &*query.assertion.clone() {
+            stmts.clone()
+        } else {
+            panic!("internal error: integer_ring")
+        };
+
+    if reqs_enss.len() != 2 {
+        // we expect exactly two block statements
+        // one for requires and one for ensures
+        panic!("internal error: integer_ring")
+    }
+
+    let reqs = if let air::ast::StmtX::Block(stmts) = &*reqs_enss[0] {
+        stmts
     } else {
         panic!("internal error: integer_ring")
     };
 
+    let enss = if let air::ast::StmtX::Block(stmts) = &*reqs_enss[1] {
+        stmts
+    } else {
+        panic!("internal error: integer_ring")
+    };
```

Comment:
This sort of code is what I'd hope to avoid through my suggestion below to use a `Query` or `StmtX` type specific to Singular.

Reply:
Yeah this part is now eliminated.